### PR TITLE
SSLEngineWebSocketServerFactory allows more customization

### PR DESCRIPTION
**Pull request by BroHammie**
## Description

Introduces a new class SSLEngineWebSocketServerFactory that takes a SSLEngine as parameter allowing for more customization.

## Related Issue

https://github.com/TooTallNate/Java-WebSocket/issues/838

## Motivation and Context

Now I can create a WebSocketServerFactory with a SSLEngine that has NeedClientAuth set to true, thus making Server request a client certificate on connect.

## How Has This Been Tested?

Created a TwoWaySSLServerExample(included in PR) that sets the value. Running wireshark with and without that flag set I can see a difference and the server sending a "client certificate request"

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/main/java/org/java_websocket/server/SSLEngineWebSocketServerFactory.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+public class SSLEngineWebSocketServerFactory implements WebSocketServerFactory {
+
+  private final SSLEngine sslEngine;
+  protected ExecutorService exec;
+
+  public SSLEngineWebSocketServerFactory(SSLEngine sslEngine) {
+    this(sslEngine, Executors.newSingleThreadScheduledExecutor());
+  }
+
+  private SSLEngineWebSocketServerFactory(SSLEngine sslEngine, ExecutorService exec) {
```
Comment (PhilipRoman) [inline src/main/java/org/java_websocket/server/SSLEngineWebSocketServerFactory.java]: Is there any specific reason why this constructor is private and not public?

File: src/main/java/org/java_websocket/server/SSLEngineWebSocketServerFactory.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  @Override
+  public ByteChannel wrapChannel(SocketChannel channel, SelectionKey key) throws IOException {
+    return new SSLSocketChannel2(channel, sslEngine, exec, key);
+  }
+
+  @Override
+  public void close() {
+
```
Comment (PhilipRoman) [inline src/main/java/org/java_websocket/server/SSLEngineWebSocketServerFactory.java]: Looks like [DefaultSSLWebSocketServerFactory](https://github.com/TooTallNate/Java-WebSocket/blob/fd4d55c2b62fa119d96f7178f47d92e3a9184de3/src/main/java/org/java_websocket/server/DefaultSSLWebSocketServerFactory.java#L85/TooTallNate/Java-WebSocket/tree/master/src/main/java/org/java_websocket/server/DefaultSSLWebSocketServerFactory.java) closes the  executor service in it's `close` method. Maybe it should be done here as well.

Comment (marci4) [reply src/main/java/org/java_websocket/server/SSLEngineWebSocketServerFactory.java]: maybe even extend?

File: src/main/java/org/java_websocket/server/SSLParametersWebSocketServerFactory.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  public SSLParametersWebSocketServerFactory(SSLContext sslContext, SSLParameters sslParameters) {
+    this(sslContext, Executors.newSingleThreadScheduledExecutor(), sslParameters);
+  }
+
+  /**
+   * New CustomSSLWebSocketServerFactory configured to only support given protocols and given cipher suites.
+   *
+   * @param sslContext          - can not be <code>null</code>
+   * @param executerService     - can not be <code>null</code>
+   * @param sslParameters    - sslParameters
```
Comment (marci4) [inline src/main/java/org/java_websocket/server/SSLParametersWebSocketServerFactory.java]: sslParameters cannot be null, could you please adjust the JavaDoc

File: src/main/java/org/java_websocket/server/SSLParametersWebSocketServerFactory.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      throw new IllegalArgumentException();
+    }
+    this.sslParameters = sslParameters;
+  }
+
+  @Override
+  public ByteChannel wrapChannel(SocketChannel channel, SelectionKey key) throws IOException {
+    SSLEngine e = sslcontext.createSSLEngine();
+    e.setUseClientMode(false);
+    if (sslParameters != null) {
```
Comment (marci4) [inline src/main/java/org/java_websocket/server/SSLParametersWebSocketServerFactory.java]: No check needed since we throw an IllegalArgumentException()

**Comment by marci4** [comment]
@BroHammie sorry for again requesting changes :(

**Comment by marci4** [comment]
@BroHammie thank you again for your contribution!

**Pull request merged**